Color picker update with team color presets

drogoganor opened the pull request:
Changes to the color picker and ColorValidator to support built-in team colors.

![colorpresets](https://user-images.githubusercontent.com/1639681/36637331-d24ae60a-1a25-11e8-8d24-799349362be3.png)

Preset colors are defined in mod.yaml as TeamColorPresets. These are based off the colors in the original games, with a couple of extras I selected myself.

Users can also save a selected color in the Custom Colors by holding LCtrl and clicking a swatch.

Bots added to skirmish or multiplayer games will use a preset color if there are any available. If all preset colors are used, a random color will be used instead.

The previous attempt at this was: https://github.com/OpenRA/OpenRA/pull/13072

drogoganor committed:
Color picker update with team color presets. Bots added will use available team color presets by default

GraionDilach commented on OpenRA.Mods.Common/Widgets/ColorPaletteSwatchWidget.cs:
```diff
@@ -0,0 +1,79 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2016 The OpenRA Developers (see AUTHORS)
```
Outdated header.

MustaphaTR commented:
> These are based off the colors in the original games, with a couple of extras I selected myself.

I think we should refer the tests done by @pi99y on #11248. Which was meant to be differ from eachother and tilesets as much as possible (he had 15 color for each mod tho, not 10). Rather than sticking to original colors.

pchote commented on OpenRA.Mods.Common/Widgets/ColorPaletteSwatchWidget.cs:
```diff
@@ -0,0 +1,79 @@
+
+namespace OpenRA.Mods.Common.Widgets
+{
+	public class ColorPaletteSwatchWidget : BackgroundWidget
```
Can you please add the click handling to `ColorBlockWidget` instead of creating a new widget type?

pchote commented on OpenRA.Game/Manifest.cs:
```diff
@@ -69,12 +69,13 @@ public class Manifest
 		public readonly string[] SoundFormats = { };
 		public readonly string[] SpriteFormats = { };
 		public readonly string[] PackageFormats = { };
+		public readonly string[] TeamColorPresets = { };
```
Please define these on the `ColorValidator` instead of creating a new engine-level field.  `IGlobalModData` implementations can automatically parse information out of the mod.yaml: see `MapGrid` for a simple example.

pchote commented on OpenRA.Mods.Common/ColorValidator.cs:
```diff
@@ -86,6 +86,62 @@ public bool IsValid(Color askedColor, out Color forbiddenColor, IEnumerable<Colo
 			return true;
 		}
 
+		public List<Color> GetModPresetColors()
```
It would be better to stick with `HSLColor`s from the start, otherwise you need to needlessly convert between types.

pchote commented on mods/common/chrome/color-picker.yaml:
```diff
@@ -42,3 +42,23 @@ Background@COLOR_CHOOSER:
 			Y: 21
 			Width: 80
 			Height: 73
+		Label@PRESETCOLORSLABEL:
```
Having everything on the one panel looks awkward IMO.
Can you put these into a separate tab instead?  Here is an old mockup from when the color picker was originally being built showing how it could look:

![htoxo14](https://user-images.githubusercontent.com/167819/36641583-b65b979c-1a29-11e8-8a1f-ca00f2b46fe3.png)

drogoganor replied:
Thanks for the feedback. I'll give this another try.

drogoganor commented:
Thanks all, I'll give this another shot soon.

drogoganor commented:
I have a new version of this in the works, here's what it looks like so far. The colors used are the ones recommended by pi99y.

![colorpresets2](https://user-images.githubusercontent.com/1639681/36726873-f2ed834e-1c06-11e8-837c-afb15681515b.png)

Any advice you have would be welcomed. I'm not so sure of how it looks. Is a Clear button even required for the custom palette? Should it prompt the user to ensure they want to delete their custom palette?

pchote commented:
The way I would do it is to make the custom colors a rolling queue and remove the Clear button.  Then, instead of using Ctrl+click, I would have an explicit "Store" button below the "Random" button on the side.

drogoganor commented:
I've changed things to be that way, with a Store button. I quite like it. I will need to do some more rework though. I also increased the size of the swatches, to have two rows of 8.

![colorpresets3](https://user-images.githubusercontent.com/1639681/36729045-93b7bef0-1c0e-11e8-96d6-d3536eb9b2dd.png)

GraionDilach commented:
I wouldn't go requiring 16 preset colors, but format the code to allow it being set up with 8 colors alone as well.

As #11248 proves, it's really hard to set up 16 significantly different colors on all terrain types of a mod. Sure, there is a setup for RA1 in the ticket, but the same color shades might not work for Tiberian Sun or the thirdparty mods (KKND, D2, RA2 comes in mind) which probably only deliver 8 colors from their respective games.

drogoganor commented:
Here's what it looks like with just 8, a random shortlist of colors:

![colorpresets4](https://user-images.githubusercontent.com/1639681/36841919-b60ed712-1d95-11e8-9a21-482b2006c7e7.png)

I've made the swatches including the label be dynamically generated so we can choose how many we want of each.

drogoganor commented:
For reference, the colors included for all 4 mods. TS is the same colors for CNC as this wasn't included in pi99y's analysis: https://github.com/OpenRA/OpenRA/issues/11248

![colorpresets5](https://user-images.githubusercontent.com/1639681/36847997-ac816c12-1dab-11e8-9af7-8291fa45a174.png)

drogoganor commented:
Here's what I have so far. Swatch is now a template item which can be resized in the yaml instead of constructed by code.

![colorpresets6](https://user-images.githubusercontent.com/1639681/36935243-2dc65fbc-1f41-11e8-9337-3063b22db056.png)

pchote commented on OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs:
```diff
@@ -65,6 +73,175 @@ public ColorPickerLogic(Widget widget, ModData modData, World world, HSLColor in
+				for (int i = 0; i < colorList.Length; i++)
+				{
+					var nameString = controlName + i;
+					var newSwatch = new ColorBlockWidget()
```
This can simplify to `newSwatch = template.Clone()`

drogoganor replied:
Is it alright if it looks like this afterward?

```
var newSwatch = (ColorBlockWidget) template.Clone();
newSwatch.IsVisible = () => true;
newSwatch.Id = nameString;
newSwatch.X = x.ToString();
newSwatch.Y = y.ToString();
newSwatch.Initialize(new WidgetArgs());
```

drogoganor commented:
New version having resized the mixer:

![colorpresets6](https://user-images.githubusercontent.com/1639681/36936129-a9e9d7da-1f4c-11e8-9ebc-4120c09d45e0.png)

pchote commented:
Looks great!

One last UI nit: Can you please make the Random and Store buttons a few px wider so that they have the same margin on the left as on the right?

drogoganor commented:
Good one on the dialog2. This is what it looks like now:

![colorpresets6](https://user-images.githubusercontent.com/1639681/36936370-668a38be-1f50-11e8-8336-99ebadc9b2f8.png)

drogoganor commented:
Changed swatch creation to use a clone of the template swatch. I had to set Visible to true and also IsVisible = () => true; to get it to work correctly. Also fixed a bug with swatch removal after moving to the new UI element types.

pchote commented on OpenRA.Mods.Common/ColorValidator.cs:
```diff
@@ -24,6 +24,7 @@ public class ColorValidator : IGlobalModData
 		public readonly int Threshold = 0x50;
 		public readonly float[] HsvSaturationRange = new[] { 0.25f, 1f };
 		public readonly float[] HsvValueRange = new[] { 0.2f, 1.0f };
+		public readonly string[] TeamColorPresets = { };
```
Likewise here.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs:
```diff
@@ -65,6 +73,174 @@ public ColorPickerLogic(Widget widget, ModData modData, World world, HSLColor in
+					}
+					else
+					{
+						// Wrap around if reached maximum
```
IMO it would feel much better ingame if this always inserted the color at the start (`list.Insert(0, hexColor)`), rolling all the older colors one step forward in the list.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs:
```diff
@@ -65,6 +73,174 @@ public ColorPickerLogic(Widget widget, ModData modData, World world, HSLColor in
+
+				int initX = x;
+
+				for (int i = 0; i < colorList.Length; i++)
```
Style nit: loop indices should be declared as `var`

pchote reviewed (commented):
Looking good!

I have one moderate structural request that should simplify the implementation a lot.  I expect to have some smaller comments once this has been done.

drogoganor commented:
Sure thing, using HSLColor was a really good idea. I didn't know it would automatically serialize like that. It simplified things a bit.

I've added the latest change as a separate commit, because it might be a bit easier to review that way. Although if I should use a single squashed commit instead, let me know. I was going to squash it into a single commit at the end once it gets approved.

pchote reviewed (commented):
Here are a few more suggestions to further simplify things:

drogoganor commented:
I've added the suggested changes and merged into a single commit. Much neater code!

drogoganor commented:
`var paletteCols = 8;
var palettePresetRows = 2;
var paletteCustomRows = 1;

MiniYaml yaml;
if (logicArgs.TryGetValue("PaletteColumns", out yaml))
	paletteCols = Convert.ToInt32(yaml.Value);
if (logicArgs.TryGetValue("PalettePresetRows", out yaml))
	palettePresetRows = Convert.ToInt32(yaml.Value);
if (logicArgs.TryGetValue("PaletteCustomRows", out yaml))
	paletteCustomRows = Convert.ToInt32(yaml.Value);`

Something like this or can it be a bit neater somehow?

pchote reviewed (commented):
A few last polish nits, then LGTM!

Can you please also apply the following yaml changes?

```diff
diff --git a/mods/cnc/chrome/color-picker.yaml b/mods/cnc/chrome/color-picker.yaml
index 6507171228..e7419c6116 100644
--- a/mods/cnc/chrome/color-picker.yaml
+++ b/mods/cnc/chrome/color-picker.yaml
@@ -5,26 +5,26 @@ Background@COLOR_CHOOSER:
 		PaletteCustomRows: 1
 	Background: panel-black
 	Width: 311
-	Height: 150
+	Height: 148
 	Children:
 		Button@RANDOM_BUTTON:
 			Key: tab
 			X: 229
-			Y: 91
+			Y: 89
 			Width: 77
 			Height: 25
 			Text: Random
 		Button@STORE_BUTTON:
 			X: 229
-			Y: 120
+			Y: 118
 			Width: 77
 			Height: 25
 			Text: Store
 			Font: Bold
 		ActorPreview@PREVIEW:
-			X: 230
+			X: 232
 			Y: 7
-			Width: 80
+			Width: 77
 			Height: 73
 			Animate: true
 		Button@MIXER_TAB_BUTTON:
@@ -41,8 +41,7 @@ Background@COLOR_CHOOSER:
 			Width: 80
 			Text: Palette
 			Font: Bold
-		Background@MIXER_TAB:
-			Background: panel-black
+		Container@MIXER_TAB:
 			X: 5
 			Y: 5
 			Width: PARENT_RIGHT - 91
@@ -66,7 +65,7 @@ Background@COLOR_CHOOSER:
 					X: 0
 					Y: 22
 					Width: PARENT_RIGHT
-					Height: 94
+					Height: 92
 					Children:
 						ColorMixer@MIXER:
 							X: 2
@@ -104,7 +103,7 @@ Background@COLOR_CHOOSER:
 							Width: PARENT_RIGHT - 4
 							Height: 58
 							X: 2
-							Y: 15
+							Y: 16
 							Children:
 								ColorBlock@COLORPRESET:
 									X: 0
@@ -117,7 +116,7 @@ Background@COLOR_CHOOSER:
 							Width: PARENT_RIGHT - 4
 							Height: 13
 							X: 2
-							Y: 72
+							Y: 71
 							Children:
 								Label@LABEL:
 									Font: TinyBold
diff --git a/mods/common/chrome/color-picker.yaml b/mods/common/chrome/color-picker.yaml
index 81e4ef7ad1..d67611fca3 100644
--- a/mods/common/chrome/color-picker.yaml
+++ b/mods/common/chrome/color-picker.yaml
@@ -5,27 +5,27 @@ Background@COLOR_CHOOSER:
 		PaletteCustomRows: 1
 	Background: dialog2
 	Width: 326
-	Height: 156
+	Height: 154
 	Children:
 		Button@RANDOM_BUTTON:
 			Key: tab
 			X: 245
-			Y: 97
+			Y: 95
 			Width: 76
 			Height: 25
 			Text: Random
 			Font: Bold
 		Button@STORE_BUTTON:
 			X: 245
-			Y: 126
+			Y: 124
 			Width: 76
 			Height: 25
 			Text: Store
 			Font: Bold
 		ActorPreview@PREVIEW:
 			X: 245
-			Y: 21
-			Width: 80
+			Y: 13
+			Width: 76
 			Height: 73
 		Button@MIXER_TAB_BUTTON:
 			X: 5
@@ -41,8 +41,7 @@ Background@COLOR_CHOOSER:
 			Width: 80
 			Text: Palette
 			Font: Bold
-		Background@MIXER_TAB:
-			Background: dialog3
+		Container@MIXER_TAB:
 			X: 5
 			Y: 5
 			Width: PARENT_RIGHT - 90
@@ -104,7 +103,7 @@ Background@COLOR_CHOOSER:
 							Width: PARENT_RIGHT - 4
 							Height: 58
 							X: 2
-							Y: 15
+							Y: 16
 							Children:
 								ColorBlock@COLORPRESET:
 									X: 0
@@ -117,7 +116,7 @@ Background@COLOR_CHOOSER:
 							Width: PARENT_RIGHT - 4
 							Height: 13
 							X: 2
-							Y: 76
+							Y: 75
 							Children:
 								Label@LABEL:
 									Font: TinyBold
diff --git a/mods/ts/chrome/color-picker.yaml b/mods/ts/chrome/color-picker.yaml
index 93f9bc184f..b833c71469 100644
--- a/mods/ts/chrome/color-picker.yaml
+++ b/mods/ts/chrome/color-picker.yaml
@@ -5,27 +5,27 @@ Background@COLOR_CHOOSER:
 		PaletteCustomRows: 1
 	Background: dialog2
 	Width: 326
-	Height: 156
+	Height: 154
 	Children:
 		Button@RANDOM_BUTTON:
 			Key: tab
 			X: 245
-			Y: 97
+			Y: 95
 			Width: 76
 			Height: 25
 			Text: Random
 			Font: Bold
 		Button@STORE_BUTTON:
 			X: 245
-			Y: 126
+			Y: 124
 			Width: 76
 			Height: 25
 			Text: Store
 			Font: Bold
 		ActorPreview@PREVIEW:
 			X: 245
-			Y: 11
-			Width: 80
+			Y: 16
+			Width: 76
 			Height: 74
 			Animate: true
 		Button@MIXER_TAB_BUTTON:
@@ -42,8 +42,7 @@ Background@COLOR_CHOOSER:
 			Width: 80
 			Text: Palette
 			Font: Bold
-		Background@MIXER_TAB:
-			Background: dialog3
+		Container@MIXER_TAB:
 			X: 5
 			Y: 5
 			Width: PARENT_RIGHT - 90
@@ -105,7 +104,7 @@ Background@COLOR_CHOOSER:
 							Width: PARENT_RIGHT - 4
 							Height: 58
 							X: 2
-							Y: 15
+							Y: 16
 							Children:
 								ColorBlock@COLORPRESET:
 									X: 0
@@ -118,7 +117,7 @@ Background@COLOR_CHOOSER:
 							Width: PARENT_RIGHT - 4
 							Height: 13
 							X: 2
-							Y: 76
+							Y: 75
 							Children:
 								Label@LABEL:
 									Font: TinyBold
```

drogoganor commented:
Thanks. I've updated the code and yaml. Also rebased to the latest bleed.

reaperrr commented:
Looks good to me, LGTM once @pchote has given this a last look-over.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs:
```diff
@@ -65,6 +73,107 @@ public ColorPickerLogic(Widget widget, ModData modData, World world, HSLColor in
+			// Setup tab controls
+			var mixerTab = widget.Get("MIXER_TAB");
+			var paletteTab = widget.Get("PALETTE_TAB");
+			var paletteTabScrollPanel = widget.Get("PALETTE_TAB_SCROLLPANEL");
```
This isn't a scrollpanel, so can you please rename to `var palettePanel` and `PALETTE_TAB_PANEL`?

pchote reviewed (dismissed):
One last minor nit, then LGTM

drogoganor commented:
Thanks for all your help!

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/57fdded66d7e3182066d3c5ac3724279b0ff55c7

pi99y commented:
Yay! Thank you guys!
